Mhd/ar samples

mhdostal opened the pull request:
Adds three AR samples to the app:  Collect data, Tabletop, and Flyover.

mhdostal commented:
@nCastle1 If you could look at the doc, that would be great.  I still need to add screenshots to the three readme files and need icon/iamge for the "Augmented Reality" category.

nCastle1 commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/README.md:
```diff
@@ -0,0 +1,62 @@
+
+There are two main approaches for identifying the physical location of tapped point:
+
+* **ArcGISARView.arScreenToLocation** - uses plane detection provided by ARKit/ARCore to determine where _in the real world_ the tapped point is.
```
I recommend removing the 'or ARCore' parts since in this context it is iOS specific. Applies to the other readmes as well.

nCastle1 replied:
Also affects line 58 below.

nCastle1 commented on arcgis-ios-sdk-samples/Augmented reality/Display scenes in tabletop AR/DisplayScenesInTabletopAR.swift:
```diff
@@ -0,0 +1,226 @@
+        }
+    }
+
+    // MARK: - Optional plane visualization
```
I had this in originally, but I wonder if the 'optional' part might be confusing. Maybe just say 'MARK: - Plane visualization'?

nCastle1 commented on arcgis-ios-sdk-samples/Augmented reality/Display scenes in tabletop AR/README.md:
```diff
@@ -0,0 +1,50 @@
+
+**Tabletop AR** is one of three main patterns for working with geographic information in augmented reality. See [Augmented reality]() in the guide for more information.
+
+This sample uses the ArcGIS Runtime Toolkit. See [Augmented reality]() in the guide to learn about the toolkit and how to add it to your app.
```
Here and in other readmes, the doc link has changed. `[Augmented reality]()` should change to `[Display scenes in augmented reality](https://developers.arcgis.com/ios/swift/latest/guide/display-scenes-in-augmented-reality.htm)`.

Link doesn't work now but will once we release.

nCastle1 commented on arcgis-ios-sdk-samples/Augmented reality/Explore scenes in flyover AR/README.md:
```diff
@@ -0,0 +1,45 @@
+
+## About the data
+
+This sample uses a sample [integrated mesh layer](https://www.arcgis.com/home/item.html?id=d4fb271d1cb747e696bb80adca8487fa) provided by [Vricon](https://www.vricon.com/). The integrated mesh layer shows an area around the US-Mexico border.
```
Don't forget to update to the new item: https://www.arcgis.com/home/item.html?id=dbc72b3ebb024c848d89a42fe6387a1b

nCastle1 commented on arcgis-ios-sdk-samples/Augmented reality/Explore scenes in flyover AR/README.md:
```diff
@@ -0,0 +1,45 @@
+
+**Flyover AR** is one of three main patterns for working with geographic information in augmented reality. See the [guide doc]() for more information.
+
+This sample uses the ArcGIS Runtime Toolkit. See [Augmented reality]() in the guide to learn about the toolkit and how to add it to your app.
```
Don't forget to update the guide doc link here.

nCastle1 commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/README.md:
```diff
@@ -0,0 +1,62 @@
+
+Note that unlike other scene samples, a basemap isn't shown most of the time, because the real world provides the context. Only while calibrating is the basemap displayed at 50% opacity, to give the user a visual reference to compare to.
+
+**Real-scale AR** is one of three main patterns for working with geographic information in augmented reality. See [Display scenes in augmented reality](https://developers.arcgis.com/ios/swift/latest/guide/display-scenes-in-augmented-reality.htm) in the guide for more information.
```
Sorry, this is super annoying, but I realized I gave a bad link. Should be `https://developers.arcgis.com/ios/swift/latest/guide/display-scenes-in-augmented-reality.htm` (swapped `swift` and `latest`)

mhdostal replied:
> https://developers.arcgis.com/ios/swift/latest/guide/display-scenes-in-augmented-reality.htm

This new link is the same as the old one.  Are you saying it should be:

`https://developers.arcgis.com/ios/latest/swift/guide/display-scenes-in-augmented-reality.htm`

nCastle1 replied:
Yeah that's right

mhdostal replied:
Ok, thanks.

philium commented:
@mhdostal I'm still reviewing this PR, but could you please add a `README.metadata.json` file for each new sample? Those files are needed for the developer site to update with the new samples. Let me know if you'd like some help.

mhdostal commented:
@philium How can I determine what the "redirect_from" url is?

philium commented:
Since these are brand new samples, there is no redirect URL, so just specify an empty array:

```json
    "redirect_from": [
    ],
```

nCastle1 commented on arcgis-ios-sdk-samples/Augmented reality/Display scenes in tabletop AR/DisplayScenesInTabletopAR.swift:
```diff
@@ -0,0 +1,239 @@
+        self.arView.translationFactor = geographicContentWidth / tableContainerWidth
+
+        // Set the origin camera
+        self.arView.originCamera = newCam
```
Just noticed that origin camera is set twice. Also, it looks like the scene parameter isn't necessary for this func; I think it was a holdover from an earlier version where the size was calculated programmatically.

nCastle1 commented on arcgis-ios-sdk-samples/Augmented reality/Display scenes in tabletop AR/DisplayScenesInTabletopAR.swift:
```diff
@@ -0,0 +1,239 @@
+        let geographicContentWidth = 800.0
+
+        // Physical width of the table area the scene will be placed on in meters
+        let tableContainerWidth = 5.0
```
I think smaller would probably be better for most uses - maybe 1 meter?

philium commented:
@mhdostal I'm unable to build this branch, I think because the submodule is pointing to a commit without the AR components. Could you update the submodule reference to an appropriate commit? Its fine if it isn't pointing to a release; we can always update it before merging to `master`. I just don't want `v.next` to get into an unbuildable state.

mhdostal commented:
> Could you update the submodule reference to an appropriate commit?

@philium Done

mhdostal commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/CollectDataAR.swift:
```diff
@@ -0,0 +1,574 @@
+    
+    /// Set whether continuous positioning is in use
+    /// Showing a heading slider is only appropriate when using local positioning
+    func setIsUsingContinuousPositioning(_ isContinuous: Bool) {
```
yes, good idea.

mhdostal commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/CollectDataAR.swift:
```diff
@@ -0,0 +1,574 @@
+    }
+
+    @IBAction func addFeature(_ sender: UIBarButtonItem) {
+        if graphicsOverlay.graphics.count < 1 {
```
Yes, done.

mhdostal commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/CollectDataAR.swift:
```diff
@@ -0,0 +1,574 @@
+            let imageHeight = CVPixelBufferGetHeight(coreVideoBuffer)
+            let imageWidth = CVPixelBufferGetWidth(coreVideoBuffer)
+            let imageRef = ciContext.createCGImage(coreImage,
+                                                   from: CGRect(x: 0, y: 0, width: imageHeight, height: imageWidth))
```
I believe so, as the image is rotated and there's a transform.  @nCastle1 can you confirm?

nCastle1 replied:
That's what I needed to do, but I only tested in portrait. Looking at what's in the feature service, it may not work right when the device is in landscape.

mhdostal replied:
The code states it assumes everything is in Portrait.  If the user wants to adjust the code so it works in both, they can do that, but I think we're covered.

mhdostal commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/CollectDataAR.storyboard:
```diff
@@ -0,0 +1,101 @@
+                                    </barButtonItem>
+                                    <barButtonItem style="plain" systemItem="flexibleSpace" id="bT6-cL-gQC"/>
+                                    <barButtonItem style="plain" id="gCU-57-zVn">
+                                        <segmentedControl key="customView" opaque="NO" contentMode="scaleToFill" contentHorizontalAlignment="left" contentVerticalAlignment="top" segmentControlStyle="bar" selectedSegmentIndex="0" id="SCJ-de-Koy">
```
I don't see that in Xcode 10.2.  I'm not sure how it can be misplaced, as it's part of the Toolbar and placed next to Flexible Space items so it's location is determined dynamically.

philium replied:
I saw this in Xcode 11.0. I'll take another look with your latest changes.

philium replied:
I'm still seeing this warning after pulling your latest changes. Since it is only visible in Xcode 11, I suspect it is related to UI changes in iOS 13.

mhdostal replied:
So are you OK with it as-is?

philium replied:
Since Xcode 11.0 is now the minimum requirement, I would recommend opening the storyboard in Xcode 11 and fixing the misplacement by updating the frames:

<img width="672" alt="image" src="https://user-images.githubusercontent.com/2257493/66334816-9a4b5700-e8ee-11e9-8371-dc916518b13a.png">

mhdostal replied:
Done.

mhdostal commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/CollectDataAR.swift:
```diff
@@ -0,0 +1,574 @@
+            presentationController.delegate = self
+            presentationController.barButtonItem = sourceButton
+            presentationController.permittedArrowDirections = [.down, .up]
+            present(controller, animated: true)
```
We don't want to present the controller outside the `if`.  The `else` is `presentAlert(message: "Error showing calibration view")`, so we don't want to present the controller after that.

mhdostal commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/CollectDataAR.swift:
```diff
@@ -0,0 +1,574 @@
+            presentationController.barButtonItem = sourceButton
+            presentationController.permittedArrowDirections = [.down, .up]
+            present(controller, animated: true)
+        } else {
```
ah, ok, got it; ignore my previous comment about not wanting it outside the `if`.

mhdostal commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/CollectDataAR.swift:
```diff
@@ -0,0 +1,574 @@
+import ArcGISToolkit
+import ArcGIS
+
+class CollectDataAR: UIViewController {
```
Looks like we always use `UIViewController`.

philium replied:
I'm sorry for not being clear; I was referring to the name of the subclass, not the name of the superclass. We include the suffix `ViewController` in the name of all the sample view controllers: `DisplayMapViewController`, `OpenSceneViewController`, etc.

mhdostal commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/CollectDataAR.swift:
```diff
@@ -0,0 +1,574 @@
+            }
+            
+            // Add the timer to the main run loop.
+            guard let timer = elevationTimer else { return }
```
Yes, cool.

mhdostal commented on arcgis-ios-sdk-samples/Augmented reality/Collect data in AR/CollectDataAR.swift:
```diff
@@ -0,0 +1,574 @@
+            } else {
+                if let featureEditResults = featureEditResults,
+                    featureEditResults.first?.completedWithErrors == false {
+                    self?.presentAlert(message: "Edits applied successfully")
```
That's what the other feature editing sample do.

philium commented on arcgis-ios-sdk-samples/Augmented reality/Explore scenes in flyover AR/README.md:
```diff
@@ -0,0 +1,45 @@
+
+## How to use the sample
+
+When you open the sample, you'll be viewing the scene from above. You can walk around, using your device as a window into the scene. Try moving vertically to get closer to the ground. Touch gestures which 
```
It looks like the end of the paragraph was cut off.

mhdostal replied:
@nCastle1 Can you provide the rest of the text?  It is cut off in the README in your fork of the samples as well.

nCastle1 replied:
Should just be:

> When you open the sample, you'll be viewing the scene from above. You can walk around, using your device as a window into the scene. Try moving vertically to get closer to the ground.

The statement about touch gestures was meant to be removed

philium commented on arcgis-ios-sdk-samples/Content Display Logic/ContentPList.plist:
```diff
@@ -1549,5 +1549,40 @@
+				<key>storyboardName</key>
+				<string>CollectDataAR</string>
+				<key>descriptionText</key>
+				<string>Tap on real-world objects to collect data</string>
```
```suggestion
				<string>Tap on real-world objects to collect data.</string>
```

philium commented on arcgis-ios-sdk-samples/Content Display Logic/ContentPList.plist:
```diff
@@ -1549,5 +1549,40 @@
+				<key>storyboardName</key>
+				<string>DisplayScenesInTabletopAR</string>
+				<key>descriptionText</key>
+				<string>Use augmented reality (AR) to pin a scene to a table or desk for easy exploration</string>
```
```suggestion
				<string>Use augmented reality (AR) to pin a scene to a table or desk for easy exploration.</string>
```

philium commented on arcgis-ios-sdk-samples/Content Display Logic/ContentPList.plist:
```diff
@@ -1549,5 +1549,40 @@
+				<key>storyboardName</key>
+				<string>ExploreScenesInFlyoverAR</string>
+				<key>descriptionText</key>
+				<string>Use augmented reality (AR) to quickly explore a scene more naturally than you could with a touch or mouse interface</string>
```
```suggestion
				<string>Use augmented reality (AR) to quickly explore a scene more naturally than you could with a touch or mouse interface.</string>
```

mhdostal commented:
> I finished reviewing your changes to Collect Data in AR and the other two samples. I had a question about the image `Augmented reality_icon`. Where did you get it? It looks like it's vector-based. Could you get it as a vector PDF?

It's a direct copy of the "Scenes" icon.  I don't know where it originally came from.

philium commented:
@mhdostal Did you see my comments about the calibration view?

mhdostal commented:
> Did you see my comments about the calibration view?

Yes.  I changed the button color to yellow, to distinguish it from the default button

To answer your specific questions:
- Calibration is applied instantly when the sliders change
- I've changed the "Calibration" button to only say "Calibration" (and not "Finish Calibrating")

So since the calibration is applied instantly, and the user can see immediately the result as they move the slider through the transparent base map, I don't think things need to change.

philium commented:
Thanks!

philium commented:
I added @saratchandrakarumuri as a reviewer so that he can validate the new samples (if he hasn't done so already).

mhdostal commented:
@saratchandrakarumuri Let me know if you find any issues.  Thanks!

dg0yal commented:
Let's get this in *now*.  The blog post announcing the release has gone live
@saratchandrakarumuri log follow on issue for any problems you find.
